lib/charms/operator_libs_linux/v0/sysctl.py

```diff
@@ -0,0 +1,241 @@
+        return "<{}.{}>".format(type(self).__module__, type(self).__name__)
+
+    @property
+    def message(self):
+        """Return the message passed as an argument."""
+        return self.args[0]
+
+
+class SysctlError(Error):
+    """Raised when there's an error running sysctl command."""
+
+
+class SysctlPermissionError(Error):
+    """Raised when there's an error applying values in sysctl."""
+
+
+class ValidationError(Error):
+    """Exception representing value validation error."""
+
+
+class SysctlConfig(Mapping[str, int]):
+    """Represents the state of the config that a charm wants to enforce."""
+
+    def __init__(self, name: Optional[str] = None) -> None:
+        self.name = name
+        self._data = self._load_data()
+
+    def __contains__(self, key: str) -> bool:
+        """Check if key is in config."""
+        return key in self._data
+
+    def __len__(self):
+        """Get size of config."""
+        return len(self._data)
+
+    def __iter__(self):
+        """Iterate over config."""
+        return iter(self._data)
+
+    def __getitem__(self, key: str) -> int:
+        """Get value for key form config."""
+        return self._data[key]
+
+    @property
+    def name(self) -> str:
+        """Name used to create the lib file."""
+        return self._name
+
+    @name.setter
+    def name(self, value: str):
+        if value is None:
+            self._name, *_ = os.getenv("JUJU_UNIT_NAME", "unknown").split("/")
+        else:
+            self._name = value
+
+    @property
+    def charm_filepath(self) -> Path:
+        """Name for resulting charm config file."""
+        return Path(f"{SYSCTL_DIRECTORY}/90-juju-{self.name}")
+
+    @property
+    def charm_config_exists(self) -> bool:
+        """Return whether the charm config file exists."""
+        return os.path.exists(self.charm_filepath)
+
+    @property
+    def merged_config_exists(self) -> bool:
+        """Return whether a merged config file exists."""
+        return os.path.exists(SYSCTL_FILENAME)
+
+    def update(self, config: Dict[str, dict]) -> None:
+        """Update sysctl config options with a desired set of config params."""
+        self._parse_config(config)
+
+        # NOTE: case where own charm calls update() more than once. Remove first so
+        # we don't get validation errors.
+        if self.charm_config_exists:
+            self.remove()
+
+        conflict = self._validate()
+        if conflict:
+            msg = f"Validation error for keys: {conflict}" 
+            raise ValidationError(msg)
+
+        snapshot = self._create_snapshot()
+        logger.debug(f"Created snapshot for keys: {snapshot}")
+        try:
+            self._apply()
+        except SysctlPermissionError:
+            self._restore_snapshot(snapshot)
+            raise
+        except SysctlError:
+            raise
+
+        self._create_charm_file()
+        self._merge()
+
+    def remove(self) -> None:
```

Comment:
The remove function should also do apply, or not?

Reply:
I don't think it _needs_ to apply. If the files exist on the charm it means that they were previously applied and passed validation

Reply:
Yes, but you are removing those configuration, so you want to remove them from applied options. So if charm-a configure `vm.swappiness=10` and no other charm was using it, we should be able to disable it on sysctl.

Reply:
Yes and no. If charm-a sets swappiness, and then its removed, what value do we set? There is no option to _reset to default_ on `sysctl`. Rebooting will reset those values to whatever the machine had, but I don't see other way around it.

Reply:
I think we can do `sysctl apply --system` and if there was `vm.swappiness` it will be reconfigured (e.g. if `vm.swappinness` is in any `10-*.cfg`. But you are right if no other config contains `vm.swappiness` it will not be reconfigured. I do not have strong opinion here, but still reconfigure it to whatever other configuration are trying to configure make sense to me. 

BTW now I realized that we should do `sysctl -p /etc/sysctl/95-juju-charm.cfg`, I know that is should be same as providing value and key in cmd, but still it would make sense to me.

Reply:
I'm doing same for GRUB and I do not see any reason why that would be a problem. At least we should mention that this function is only removing config and not applying anything.